### PR TITLE
refactor ProcessFactory to help debugging

### PR DESCRIPTION
**Pull request by xu-chiheng**

### CODE REVIEW

File: core/org.eclipse.cdt.core.native/src/org/eclipse/cdt/utils/spawner/ProcessFactory.java [hunk 1]
```diff
@@ -81,8 +176,74 @@ public Builder gracefulExitTimeMs(int gracefulExitTimeMs) {
...
+			/*
+			 * set breakpoint on next line to inspect sb when debugging, to see the
+			 * ultimate parameters of ProcessBuilder
+			 */
+			return sb;
+		}
+
 		public Process start() throws IOException {
 			cmdarray = modifyCmdArrayIfFlatpak(cmdarray);
+			debug();
```
Comment (jonahgraham) [inline core/org.eclipse.cdt.core.native/src/org/eclipse/cdt/utils/spawner/ProcessFactory.java:246]: I am not keen on having this run all the time. Can you consider putting this behind a trace flag or simply commenting out this line?

```suggestion
			// Uncomment the next line, step over debug() to get detailed info on what is being launched.
			// var debug = debug();
```

File: core/org.eclipse.cdt.core.native/src/org/eclipse/cdt/utils/spawner/ProcessFactory.java [hunk 2]
```diff
@@ -33,6 +40,94 @@ public class ProcessFactory {
...
+				} else {
+					// removed
+					sb.append('-');
+					sb.append(key);
+					sb.append('=');
+					sb.append(valueA);
+					sb.append('\n');
+				}
+			} else {
+				// environmentB.containsKey(key)
```
Comment (jonahgraham) [inline core/org.eclipse.cdt.core.native/src/org/eclipse/cdt/utils/spawner/ProcessFactory.java:119]: Remove commented out code.
```suggestion
```

**Review by jonahgraham — CHANGES_REQUESTED** [review]

**Comment by xu-chiheng** [comment]
On Tue, Dec 6, 2022 at 4:07 AM Jonah Graham ***@***.***>
wrote:

> ***@***.**** requested changes on this pull request.
> ------------------------------
>
> In
> core/org.eclipse.cdt.core.native/src/org/eclipse/cdt/utils/spawner/ProcessFactory.java
> <https://github.com/eclipse-cdt/cdt/pull/191#discussion_r1039008309>:
>
> > @@ -33,6 +40,94 @@ public class ProcessFactory {
>
>  	private boolean hasSpawner;
>
>  	private Runtime runtime;
>
>
>
> +	private static TreeMap<String, String> newEmptyEnvironment() {
>
> +		TreeMap<String, String> environment;
>
> +		if (Platform.getOS().equals(Platform.OS_WIN32)) {
>
>
> What is the purpose of case insensitive comparator only on windows? I
> assume it is to make Path and PATH appear next to each other when
> iterating? However it will leave both Path and PATH in the map which won’t
> represent what happens in the process
>
Wrong, the purpose is to treat Path and PATH as the same key.


> ------------------------------
>
> In
> core/org.eclipse.cdt.core.native/src/org/eclipse/cdt/utils/spawner/ProcessFactory.java
> <https://github.com/eclipse-cdt/cdt/pull/191#discussion_r1040039922>:
>
> >  		public Process start() throws IOException {
>
>  			cmdarray = modifyCmdArrayIfFlatpak(cmdarray);
>
> +			debug();
>
>
> I am not keen on having this run all the time. Can you consider putting
> this behind a trace flag or simply commenting out this line?
> ⬇️ Suggested change
>
> -			debug();
>
> +			// Uncomment the next line, step over debug() to get detailed info on what is being launched.
>
> +			// var debug = debug();
>
> can be done, comment it out.

> ------------------------------
>
> In
> core/org.eclipse.cdt.core.native/src/org/eclipse/cdt/utils/spawner/ProcessFactory.java
> <https://github.com/eclipse-cdt/cdt/pull/191#discussion_r1040040360>:
>
> > +						sb.append('+');
>
> +						sb.append(key);
>
> +						sb.append('=');
>
> +						sb.append(valueB);
>
> +						sb.append('\n');
>
> +					}
>
> +				} else {
>
> +					// removed
>
> +					sb.append('-');
>
> +					sb.append(key);
>
> +					sb.append('=');
>
> +					sb.append(valueA);
>
> +					sb.append('\n');
>
> +				}
>
> +			} else {
>
> +				// environmentB.containsKey(key)
>
>
> Remove commented out code.
> ⬇️ Suggested change
>
> -				// environmentB.containsKey(key)
>
> This is just a comment for better reading the code .

File: core/org.eclipse.cdt.core.native/src/org/eclipse/cdt/utils/spawner/ProcessFactory.java [hunk 3]
```diff
@@ -33,6 +40,94 @@ public class ProcessFactory {
 	private boolean hasSpawner;
 	private Runtime runtime;
 
+	private static TreeMap<String, String> newEmptyEnvironment() {
+		TreeMap<String, String> environment;
+		if (Platform.getOS().equals(Platform.OS_WIN32)) {
```
[the remark this answers is not in the document]
Comment (jonahgraham) [reply core/org.eclipse.cdt.core.native/src/org/eclipse/cdt/utils/spawner/ProcessFactory.java:45]: Please disregard my incorrect comment here.

**Review by jonahgraham — APPROVED** [review]
LGTM, let me know when you are finished pushing commits to this PR.

**Comment by xu-chiheng** [comment]
On Tue, Dec 6, 2022 at 6:07 AM Jonah Graham ***@***.***>
wrote:

> ***@***.**** approved this pull request.
>
> LGTM, let me know when you are finished pushing commits to this PR.
>
already finished.

**Comment by jonahgraham** [comment]
On Tue, 6 Dec 2022 at 13:02, 徐持恒 Xu Chiheng ***@***.***>
wrote:

> On Tue, Dec 6, 2022 at 6:07 AM Jonah Graham ***@***.***>
> wrote:
>
> > ***@***.**** approved this pull request.
> >
> > LGTM, let me know when you are finished pushing commits to this PR.
> >
> already finished.
>

😀

**Pull request merged**